pageserver/src/walredo.rs

```diff
@@ -319,130 +413,63 @@ impl PostgresRedoManager {
-                // it could happen that if walredo always fails immediately, we spawn processes faster
-                // than we can SIGKILL & `wait` for them to exit. By doing it the way we do here,
-                // we limit this risk of run-away to at most $num_runtimes * $num_executor_threads.
-                // This probably needs revisiting at some later point.
-                match self.redo_process.get() {
-                    None => (),
-                    Some(guard) => {
-                        match &*guard {
-                            ProcessOnceCell::ManagerShutDown => {}
-                            ProcessOnceCell::Spawned(guard_proc) => {
-                                if Arc::ptr_eq(&proc, guard_proc) {
-                                    // We're the first to observe an error from `proc`, it's our job to take it out of rotation.
-                                    guard.take_and_deinit();
-                                } else {
-                                    // Another task already spawned another redo process (further up in this method)
-                                    // and put it into `redo_process`. Do nothing, our view of the world is behind.
-                                }
-                            }
+                let len = records.len();
+                let nbytes = records.iter().fold(0, |acumulator, record| {
+                    acumulator
+                        + match &record.1 {
+                            NeonWalRecord::Postgres { rec, .. } => rec.len(),
+                            _ => unreachable!("Only PostgreSQL records are accepted in this batch"),
                         }
-                    }
+                });
+
+                WAL_REDO_TIME.observe(duration.as_secs_f64());
+                WAL_REDO_RECORDS_HISTOGRAM.observe(len as f64);
+                WAL_REDO_BYTES_HISTOGRAM.observe(nbytes as f64);
+
+                debug!(
+                        "postgres applied {} WAL records ({} bytes) in {} us to reconstruct page image at LSN {}",
+                        len,
+                        nbytes,
+                        duration.as_micros(),
+                        lsn
+                    );
+
+                if let Err(e) = result.as_ref() {
+                    error!(
+                            "error applying {} WAL records {}..{} ({} bytes) to key {key}, from base image with LSN {} to reconstruct page image at LSN {} n_attempts={}: {:?}",
+                            records.len(),
+                            records.first().map(|p| p.0).unwrap_or(Lsn(0)),
+                            records.last().map(|p| p.0).unwrap_or(Lsn(0)),
+                            nbytes,
+                            base_img_lsn,
+                            lsn,
+                            n_attempts,
+                            e,
+                        );
```

Comment:
Let's fix the indetation, I was about to fire a PR for the stuff I fixed from this file.

This PR makes things worse.